server/container/util/src/main/java/org/apache/james/util/FluentFutureStream.java

```diff
@@ -153,6 +175,12 @@ private FluentFutureStream(CompletableFuture<Stream<T>> completableFuture) {
             .thenApply(stream -> stream.filter(predicate)));
     }
 
+    public FluentFutureStream<T> thenFilter(Function<T, CompletableFuture<Boolean>> futurePredicate) {
+        return thenFlatComposeOnOptional(t -> futurePredicate.apply(t)
```

Comment:
I understand that thenFlatComposeOnOptional already existed but it's becoming ridiculous. Are you going to implement a method for every single monad combinations ? What about using usual FP technics (like monad transformer) instead ?

Reply:
I did try hard to make things more composable here: https://github.com/linagora/james-project/pull/1740